### ISSUE TITLE
KubernetesReactiveDiscoveryClient cache configuration issues

### ISSUE THREAD

**Issue by habelson** [spring-cloud/spring-cloud-kubernetes#1641]
**Describe the bug**
Spring Cloud 2023.0.1 (and prior)

We have an ecosystem of spring boot microservices, some are reactive and some are blocking. Some have their own need for caching and some don't. We have run into an issue with the presence of the @Cacheable annotation in `KubernetesReactiveDiscoveryClient` in reactive microservices that have `@EnableCaching` for their own purposes.

This has surfaced as several issues:

1. `DefaultKubernetesServiceInstance` is not `Serializable`, preventing some caching implementations from functioning correctly by default
2. JSON Serialization of  `DefaultKubernetesServiceInstance` fails by default because the properties discovered by the getters don't match whats available in the constructor. This requires configuring the ObjectMapper to ignore unknown properties.
3. With `spring.cloud.kubernetes.loadbalancer.mode=POD` the cached value of the discovered services is the POD address which changes on redeploy, and there is nothing setup to invalidate the cache.
4. The blocking implementation of KubernetesDiscoveryClient does not have the same use of @Cacheable and therefore has different behavior from its reactive counterpart 

While I'm not convinced there is necessarily a bug here, I am concerned on the lack of documentation.  What sort of cache configuration is expected to be in place (backend library, TTL, etc) for the reactive discovery client to behave as intended? Additionally, was the caching behavior in the reactive version of the class intended to be different from the blocking version?

**Comment by ryanjbaxter** [comment]
Just trying to understand exactly what you are trying to achieve with caching…

are you trying to cache the services returned by the discovery client so it doesn’t need to make a request to the api server/discovery server?

**Comment by ryanjbaxter** [comment]
I don't think there is a good solution that doesn't cause someone some amount of pain, its unfortunate that one of the clients already has `@Cacheable`.  For the 2024 release train I would say that we just make it possible for a user to cache the services if they want.  That might mean creating their own bean with the `@Cacheable` annotation if they are using one of the clients that doesn't have the annotation.  Then in 2025 we can correct this and break functionality.

**Comment by ryanjbaxter** [comment]
Yes, but also I would like to explore if we need to make any changes/fixes to the code base today to all things to be cached by using custom bean definitions

**Comment by ryanjbaxter** [comment]
No what I am concerned about is that there is something in the way we have implemented the discovery client that would prevent users from providing their own caching implementation

**Comment by habelson** [comment]
I am curious if there is any movement on this since this discussion last year? Is it still planned to be addressed in the 2025 release train?

**Comment by wind57** [comment]
let's start with the first two points here that you have raised since they are probably the simplest to address.

> DefaultKubernetesServiceInstance is not Serializable, preventing some caching implementations from functioning correctly by default

You are correct and I see no reason not to make them `Serializable`. That is easy to do. 

> JSON Serialization of DefaultKubernetesServiceInstance fails by default because the properties discovered by the getters don't match whats available in the constructor. This requires configuring the ObjectMapper to ignore unknown properties.

I'm not sure I follow you here, but may be because we have done some refactoring. 

What I mean, is that I wrote some tests and they all pass. Can you have a look please and let me know if that was your concern? 

https://github.com/spring-cloud/spring-cloud-kubernetes/pull/2062/files

**Comment by wind57** [comment]
this is not completed though :( there are things pending...

**Comment by ryanjbaxter** [comment]
Sorry GitHub automatically closed it because I linked the PR
